[NCCL] warn when barrier guesses device to use

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54991 [NCCL] warn when barrier guesses device to use**

Actual proposed fix is in
https://github.com/pytorch/pytorch/pull/53934, in the meantime, would be useful
to include this LOG when barrier does not know what devices to use, and suggest
the workaround of passing in device_ids into barrier().

Differential Revision: [D27444917](https://our.internmc.facebook.com/intern/diff/D27444917/)

rohan-varma committed:
[NCCL] Log when barrier guesses device to use

Actual proposed fix is in
https://github.com/pytorch/pytorch/pull/53934, in the meantime, would be useful
to include this LOG when barrier does not know what devices to use, and suggest
the workaround of passing in device_ids into barrier().

Differential Revision: [D27444917](https://our.internmc.facebook.com/intern/diff/D27444917/)

[ghstack-poisoned]

SciPioneer commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1413,6 +1413,15 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::barrier(
+      " using best-guess GPU ",
+      deviceIdx,
+      " to perform barrier as devices used by this process are currently unknown. ",
+      "Specify device_ids in barrier() to force use of a particular device."
```
Better to say "To avoid any potential hang due to a wrong guess, please specify ...". This can improve the debuggability by mentioning the potential outcome if the warning here is ignored.

wayi1 reviewed (approved):
Thanks for improving the debuggability!

SciPioneer commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1413,6 +1413,15 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::barrier(
+      " to perform barrier as devices used by this process are currently unknown. ",
+      "Specify device_ids in barrier() to force use of a particular device."
+    );
+    LOG(WARNING) << bestGuessWarning;
```
Nit: Since `bestGuessWarning` wont' be used anywhere else, can just log everything directly after `LOG(WARNING)` and no need to construct a separate string.

rohan-varma committed:
Update on "[NCCL] warn when barrier guesses device to use"


Actual proposed fix is in
https://github.com/pytorch/pytorch/pull/53934, in the meantime, would be useful
to include this LOG when barrier does not know what devices to use, and suggest
the workaround of passing in device_ids into barrier().

Differential Revision: [D27444917](https://our.internmc.facebook.com/intern/diff/D27444917/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[NCCL] warn when barrier guesses device to use"


Actual proposed fix is in
https://github.com/pytorch/pytorch/pull/53934, in the meantime, would be useful
to include this LOG when barrier does not know what devices to use, and suggest
the workaround of passing in device_ids into barrier().

Differential Revision: [D27444917](https://our.internmc.facebook.com/intern/diff/D27444917/)

[ghstack-poisoned]
